WarningsNotAsErrors proposal (#12134)
NuGet/Home: proposed/2022/WarningsNotAsErrors.md

This feature is supported by the compiler, and we want to provide parity for both pack and restore.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
The proposed behavior here matches the compiler behavior. All the scenarios are best represented in a table.

The precedence order for the 3 properties by NuGet can summarized as:
`NoWarn > WarningsAsErrors > WarningsNotAsErrors`.

For the following table assume `NU1603` is being raised during restore.

| TreatWarningsAsErrors | NoWarn | WarningsAsErrors | WarningsNotAsErrors | Behavior | Explanation |
|-----------------------|--------|------------------|---------------------|----------|-------------|
| true | | | | All warnings are treated as errors | |

> | true | NU1603 | |  | Succeeds with no warning. | NU1603 is raised, but because of no warn it won't be shown. |

Reviewer: ```suggestion
| true | NU1603 | |  | Succeeds with no warning. | NoWarn prevents NU1603 from being raised. |
```

I get that internally, NuGet will still do the NU1603 analysis, create the internal warning, and it's the "output" code that will prevent it from being shown to the user, but I think this implementation detail just doesn't matter to customers. Customers won't see the NU1603, and I think that's all that matters to a design document.

Author: Your suggestion makes sense. 
I think it was a late update that led to the sloppy explanation.

This feature is supported by the compiler, and we want to provide parity for both pack and restore.

## Explanation

### Functional explanation

<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
The proposed behavior here matches the compiler behavior. All the scenarios are best represented in a table.

The precedence order for the 3 properties by NuGet can summarized as:
`NoWarn > WarningsAsErrors > WarningsNotAsErrors`.

For the following table assume `NU1603` is being raised during restore.

| TreatWarningsAsErrors | NoWarn | WarningsAsErrors | WarningsNotAsErrors | Behavior | Explanation |
|-----------------------|--------|------------------|---------------------|----------|-------------|
| true | | | | All warnings are treated as errors | |

> | true | | | | All warnings are treated as errors | |

Reviewer: These two rows are duplicates

> | true | | NU1603 |  | Succeeds with no warning. | NU1603 is raised, but because of no warn it won't be shown. |

Reviewer: The NoWarn column is empty, but the explanation says that "because of no warn it won't be shown"

Author: Fixed.
